### PR TITLE
feat: add frame_image, for emitting low-res frame screenshots

### PR DESCRIPTION
**Pull request by colinmarc**
Leaving this as a draft since I haven't had a chance to test it.

### CODE REVIEW

File: tracy-client/src/frame.rs [hunk 1]
```diff
@@ -114,6 +114,30 @@ impl Client {
         }
         Frame(self.clone(), name)
     }
+
+    /// Emits an image of the current frame.
```
Comment (nagisa) [inline tracy-client/src/frame.rs:118]:
```suggestion
    /// Emits an image of a frame.
```

Given that you can specify an offset to associate the image with an older frame.

File: tracy-client/src/frame.rs [hunk 2]
```diff
@@ -114,6 +114,30 @@ impl Client {
...
+    ///
+    /// The following restrictions apply:
+    /// - The image must be in RGBA format.
+    /// - The image width and height must be divisible by four.
+    /// - The image data must be less than 256 KB, and should be as small as
+    ///   possible.
+    ///
+    /// The offset indicates how many frames in the past the image was captured.
+    /// For example, an offset of 1 would associate the image with the previous
+    /// call to [frame_mark].
```
Comment (nagisa) [inline tracy-client/src/frame.rs:128]:
```suggestion
    /// call to [`frame_mark`](Client::frame_mark).
```

probably...

**Review by nagisa — APPROVED** [review]

File: tracy-client/src/frame.rs (hunk 2 above)
Comment (colinmarc) [reply tracy-client/src/frame.rs:128]: The link works without this change :shrug:

Comment (colinmarc) [reply tracy-client/src/frame.rs:128]: Oh, I see, this is actually different.

**Comment by Dinnerbone** [comment]
I've tested this and can confirm it works perfectly!

Here's a wgpu implementation:
```rust
#[derive(Debug)]
pub struct FrameCaptures {
    available_buffers: Vec<PooledBuffer>,
    submitted_captures: Vec<PendingCapture>,
    mapped_captures: Vec<PendingCapture>,
    current_frame: u64,
    blit_pipeline: wgpu::RenderPipeline,
    blit_bind_group_layout: wgpu::BindGroupLayout,
    target: Option<CaptureTarget>,
    sampler: wgpu::Sampler,
}

#[derive(Debug)]
struct PooledBuffer {
    texture: wgpu::Texture,
    view: wgpu::TextureView,
    buffer: wgpu::Buffer,
}

#[derive(Debug)]
struct CaptureTarget {
    bind_group: wgpu::BindGroup,
    width: u32,
    height: u32,
    padded_bytes_per_row: u32,
}

#[derive(Debug)]
struct PendingCapture {
    buffer: PooledBuffer,
    width: u32,
    height: u32,
    frame_num: u64,
    ready: Arc<AtomicBool>,
}

impl PendingCapture {
    pub fn upload(&self, current_frame: u64) {
        // If we're so far behind we couldn't even fit into a u8, just drop it.
        let Some(offset) = current_frame
            .checked_sub(self.frame_num)
            .and_then(|o| u8::try_from(o).ok())
        else {
            return;
        };

        // Tracy needs a raw rgba image with no padding, so we have to remove it
        let mut unpadded = Vec::with_capacity(self.width as usize * self.height as usize * 4);
        let padded_width = wgpu::util::align_to(self.width * 4, wgpu::COPY_BYTES_PER_ROW_ALIGNMENT);
        let buffer = self.buffer.buffer.slice(..).get_mapped_range();
        for row in 0..self.height {
            let start = row as usize * padded_width as usize;
            let end = start + (self.width as usize * 4);
            unpadded.extend_from_slice(&buffer[start..end]);
        }
        frame_image(
            &unpadded,
            self.width as u16,
            self.height as u16,
            offset,
            false,
        );
    }
}

impl PooledBuffer {
    pub fn new(device: &wgpu::Device, target: &CaptureTarget) -> Self {
        let texture = device.create_texture(&wgpu::TextureDescriptor {
            label: None,
            size: wgpu::Extent3d {
                width: target.width,
                height: target.height,
                depth_or_array_layers: 1,
            },
            mip_level_count: 1,
            sample_count: 1,
            dimension: wgpu::TextureDimension::D2,
            format: wgpu::TextureFormat::Rgba8Unorm,
            usage: wgpu::TextureUsages::RENDER_ATTACHMENT | wgpu::TextureUsages::COPY_SRC,
            view_formats: &[],
        });
        let view = texture.create_view(&wgpu::TextureViewDescriptor::default());
        let buffer = device.create_buffer(&wgpu::BufferDescriptor {
            label: None,
            size: (target.padded_bytes_per_row * target.height * 4) as wgpu::BufferAddress,
            usage: wgpu::BufferUsages::COPY_DST | wgpu::BufferUsages::MAP_READ,
            mapped_at_creation: false,
        });
        PooledBuffer {
            texture,
            view,
            buffer,
        }
    }
}

impl FrameCaptures {
    /// Maximum width of a texture to capture
    /// This MUST be a multiple of 4.
    const MAX_WIDTH: u32 = 320;

    /// Maximum height of a texture to capture.
    /// This MUST be a multiple of 4.
    const MAX_HEIGHT: u32 = 180;

    pub fn new(device: &wgpu::Device) -> Self {
        let blit_bind_group_layout =
            device.create_bind_group_layout(&wgpu::BindGroupLayoutDescriptor {
                label: None,
                entries: &[
                    wgpu::BindGroupLayoutEntry {
                        binding: 0,
                        visibility: wgpu::ShaderStages::FRAGMENT,
                        ty: wgpu::BindingType::Sampler(wgpu::SamplerBindingType::Filtering),
                        count: None,
                    },
                    wgpu::BindGroupLayoutEntry {
                        binding: 1,
                        visibility: wgpu::ShaderStages::FRAGMENT,
                        ty: wgpu::BindingType::Texture {
                            sample_type: wgpu::TextureSampleType::Float { filterable: true },
                            view_dimension: wgpu::TextureViewDimension::D2,
                            multisampled: false,
                        },
                        count: None,
                    },
                ],
            });
        let blit_pipeline_layout = device.create_pipeline_layout(&wgpu::PipelineLayoutDescriptor {
            label: None,
            bind_group_layouts: &[&blit_bind_group_layout],
            push_constant_ranges: &[],
        });
        let shader_source = r"
@group(0) @binding(0) var frame_sampler: sampler;
@group(0) @binding(1) var frame_texture: texture_2d<f32>;
struct VertexOutput {
    @builtin(position) position: vec4<f32>,
    @location(0) tex_coord: vec2<f32>,
};
@vertex fn vs_main(@builtin(vertex_index) vertex_index: u32) -> VertexOutput {
    var result: VertexOutput;
    let x = i32(vertex_index) / 2;
    let y = i32(vertex_index) & 1;
    let tc = vec2<f32>(
        f32(x) * 2.0,
        f32(y) * 2.0
    );
    result.position = vec4<f32>(
        tc.x * 2.0 - 1.0,
        1.0 - tc.y * 2.0,
        0.0, 1.0
    );
    result.tex_coord = tc;
    return result;
}
@fragment fn fs_main(in: VertexOutput) -> @location(0) vec4<f32> {
    return textureSample(frame_texture, frame_sampler, in.tex_coord);
}
";
        let shader_module = device.create_shader_module(wgpu::ShaderModuleDescriptor {
            label: None,
            source: wgpu::ShaderSource::Wgsl(Cow::Borrowed(shader_source)),
        });
        let blit_pipeline = device.create_render_pipeline(&wgpu::RenderPipelineDescriptor {
            label: None,
            layout: Some(&blit_pipeline_layout),
            vertex: wgpu::VertexState {
                module: &shader_module,
                entry_point: None,
                compilation_options: Default::default(),
                buffers: &[],
            },
            primitive: Default::default(),
            depth_stencil: None,
            multisample: Default::default(),
            fragment: Some(wgpu::FragmentState {
                module: &shader_module,
                entry_point: None,
                compilation_options: Default::default(),
                targets: &[Some(wgpu::TextureFormat::Rgba8Unorm.into())],
            }),
            multiview: None,
            cache: None,
        });
        let sampler = device.create_sampler(&wgpu::SamplerDescriptor {
            address_mode_u: wgpu::AddressMode::ClampToEdge,
            address_mode_v: wgpu::AddressMode::ClampToEdge,
            mag_filter: wgpu::FilterMode::Linear,
            min_filter: wgpu::FilterMode::Linear,
            ..Default::default()
        });
        Self {
            current_frame: 0,
            blit_pipeline,
            available_buffers: Default::default(),
            submitted_captures: Default::default(),
            mapped_captures: Default::default(),
            blit_bind_group_layout,
            target: None,
            sampler,
        }
    }

    pub fn set_target(&mut self, device: &wgpu::Device, texture: Option<&wgpu::Texture>) {
        if let Some(texture) = texture {
            // The ideal size of a captured texture should be 320x180
            // Preserve the texture aspect ratio (if possible) whilst maintaining that limit
            // However - the final size MUST be divisible by 4.
            let scale_x = texture.width() as f32 / Self::MAX_WIDTH as f32;
            let scale_y = texture.height() as f32 / Self::MAX_HEIGHT as f32;
            let scale = scale_x.max(scale_y).max(1.0);
            let width = (texture.width() as f32 / scale).floor() as u32 / 4 * 4;
            let height = (texture.height() as f32 / scale).floor() as u32 / 4 * 4;
            let view = texture.create_view(&Default::default());
            let bind_group = device.create_bind_group(&wgpu::BindGroupDescriptor {
                label: None,
                layout: &self.blit_bind_group_layout,
                entries: &[
                    wgpu::BindGroupEntry {
                        binding: 0,
                        resource: wgpu::BindingResource::Sampler(&self.sampler),
                    },
                    wgpu::BindGroupEntry {
                        binding: 1,
                        resource: wgpu::BindingResource::TextureView(&view),
                    },
                ],
            });

            // 4 bytes per pixel (Tracy wants RGBA)
            let bytes_per_row = width * 4;
            let padded_bytes_per_row =
                wgpu::util::align_to(bytes_per_row, wgpu::COPY_BYTES_PER_ROW_ALIGNMENT);

            if let Some(old_target) = &self.target {
                if old_target.width != width || old_target.height != height {
                    // We've changed the size of the ideal texture, let's recreate our buffers
                    self.available_buffers.clear();
                }
            }

            self.target = Some(CaptureTarget {
                bind_group,
                width,
                height,
                padded_bytes_per_row,
            });
            self.available_buffers.clear();
        } else {
            self.target = None;
        }
    }

    pub fn capture_frame(&mut self, device: &wgpu::Device, encoder: &mut wgpu::CommandEncoder) {
        let Some(target) = &self.target else {
            return;
        };

        // Take an existing buffer, or make one if we've run out
        let buffer = self
            .available_buffers
            .pop()
            .unwrap_or_else(|| PooledBuffer::new(device, target));

        // Copy the frame to the texture
        let mut render_pass = encoder.begin_render_pass(&wgpu::RenderPassDescriptor {
            label: None,
            color_attachments: &[Some(wgpu::RenderPassColorAttachment {
                view: &buffer.view,
                resolve_target: None,
                ops: Default::default(),
            })],
            depth_stencil_attachment: None,
            timestamp_writes: None,
            occlusion_query_set: None,
        });
        render_pass.set_pipeline(&self.blit_pipeline);
        render_pass.set_bind_group(0, &target.bind_group, &[]);
        render_pass.draw(0..6, 0..1);
        drop(render_pass);

        // Copy the texture to the buffer
        encoder.copy_texture_to_buffer(
            wgpu::ImageCopyTexture {
                texture: &buffer.texture,
                mip_level: 0,
                origin: wgpu::Origin3d::ZERO,
                aspect: wgpu::TextureAspect::All,
            },
            wgpu::ImageCopyBuffer {
                buffer: &buffer.buffer,
                layout: wgpu::ImageDataLayout {
                    offset: 0,
                    bytes_per_row: Some(target.padded_bytes_per_row),
                    rows_per_image: Some(target.height),
                },
            },
            wgpu::Extent3d {
                width: target.width,
                height: target.height,
                depth_or_array_layers: 1,
            },
        );
        let capture = PendingCapture {
            buffer,
            width: target.width,
            height: target.height,
            frame_num: self.current_frame,
            ready: Arc::new(AtomicBool::new(false)),
        };
        self.submitted_captures.push(capture);
    }

    pub fn finish_frame(&mut self) {
        frame_mark();
        self.current_frame += 1;
        let current_size = self.target.as_ref().map(|t| (t.width, t.height));
        for capture in std::mem::take(&mut self.submitted_captures) {
            let ready = capture.ready.clone();
            capture
                .buffer
                .buffer
                .slice(..)
                .map_async(wgpu::MapMode::Read, move |_| {
                    ready.store(true, Ordering::Relaxed);
                });
            self.mapped_captures.push(capture);
        }
        for capture in std::mem::take(&mut self.mapped_captures) {
            if capture.ready.load(Ordering::Relaxed) {
                capture.upload(self.current_frame);
                capture.buffer.buffer.unmap();

                if current_size == Some((capture.width, capture.height)) {
                    self.available_buffers.push(capture.buffer);
                }
            } else {
                self.mapped_captures.push(capture);
            }
        }
    }
}
```

**Comment by colinmarc** [comment]
Rebased!

**Comment by nagisa** [comment]
@Dinnerbone would you mind making this into a self-contained runnable example and submitting a PR for it? Something that captures plain triangles or something of the sort would be plenty awesome!

**Comment by Dinnerbone** [comment]
I took a quick look but there's a snag; we can't depend on wgpu (including in an `--example`) or it's a circular dependency, since wgpu depends on tracy.
I could make it a separate crate living in e.g. `examples/wgpu-frame-images`, if that's okay?

**Comment by nagisa** [comment]
I'd say we introduce a new crate for the time being with `publish = false`, yeah.

**Comment by colinmarc** [comment]
Does it work to just disable the `tracy` feature of wgpu?

**Pull request merged**